recipe-client-addon/bootstrap.js

```diff
@@ -29,69 +38,146 @@ const DEFAULT_PREFS = {
   "app.shield.optoutstudies.enabled": AppConstants.MOZ_DATA_REPORTING,
 };
 
-this.install = function() {};
-
-this.startup = function() {
-  // Initialize preference defaults before anything else happens.
-  const prefBranch = Services.prefs.getDefaultBranch("");
-  for (const [name, value] of Object.entries(DEFAULT_PREFS)) {
-    switch (typeof value) {
-      case "string":
-        prefBranch.setCharPref(name, value);
-        break;
-      case "number":
-        prefBranch.setIntPref(name, value);
-        break;
-      case "boolean":
-        prefBranch.setBoolPref(name, value);
-        break;
-      default:
-        throw new Error(`Invalid default preference type ${typeof value}`);
+// Logging
+const log = Log.repository.getLogger(BOOTSTRAP_LOGGER_NAME);
+log.addAppender(new Log.ConsoleAppender(new Log.BasicFormatter()));
+log.level = Services.prefs.getIntPref(PREF_LOGGING_LEVEL, Log.Level.Warn);
+
+this.Bootstrap = {
+  initShieldPrefs(defaultPrefs) {
+    const prefBranch = Services.prefs.getDefaultBranch("");
+    for (const [name, value] of Object.entries(defaultPrefs)) {
+      switch (typeof value) {
+        case "string":
+          prefBranch.setCharPref(name, value);
+          break;
+        case "number":
+          prefBranch.setIntPref(name, value);
+          break;
+        case "boolean":
+          prefBranch.setBoolPref(name, value);
+          break;
+        default:
+          throw new Error(`Invalid default preference type ${typeof value}`);
+      }
     }
-  }
+  },
 
-  ShieldRecipeClient.startup();
-};
+  initExperimentPrefs() {
+    const defaultBranch = Services.prefs.getDefaultBranch("");
+    const experimentBranch = Services.prefs.getBranch(STARTUP_EXPERIMENT_PREFS_BRANCH);
+
+    for (const childPrefName of experimentBranch.getChildList("")) {
+      const realPrefName = childPrefName.slice(1); // Remove leading dot
+      const realPrefType = defaultBranch.getPrefType(childPrefName);
+      const experimentPrefType = experimentBranch.getPrefType(childPrefName);
+
+      // TODO: This never passes?
+      if (realPrefType !== Services.prefs.PREF_INVALID && realPrefType !== experimentPrefType) {
+        log.error(`Error setting startup pref ${childPrefName}; pref type does not match.`);
+        continue;
+      }
+
+      switch (experimentPrefType) {
+        case Services.prefs.PREF_STRING:
+          defaultBranch.setCharPref(realPrefName, experimentBranch.getCharPref(childPrefName));
```

Comment:
So, this only sets on the default branch. Are we no longer supporting pref experiments on the user branch at all? Do we need to encode that bit into the experiment branch (with "user." or "default." prefixes or something) ?

Reply:
User prefs persist between restarts, so we don't need to set them on startup. They get set when the experiment starts.